fenwick/src/main/java/org/opencadc/fenwick/StorageSiteSync.java

```diff
@@ -0,0 +1,204 @@
+
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.Artifact;
+import org.opencadc.inventory.InventoryUtil;
+import org.opencadc.inventory.SiteLocation;
+import org.opencadc.inventory.StorageSite;
+import org.opencadc.inventory.db.ObsoleteStorageLocation;
+import org.opencadc.inventory.db.ObsoleteStorageLocationDAO;
+import org.opencadc.inventory.db.StorageSiteDAO;
+import org.opencadc.tap.TapClient;
+
+
+/**
+ * Class to pull a single StorageSite from a remote TAP (Luskan) service.  This class is enabled by the
+ * org.opencadc.fenwick.trackSiteLocations property in the fenwick.properties file.
+ */
+public class StorageSiteSync {
+    private static final Logger log = Logger.getLogger(StorageSiteSync.class);
+
+    private static final String STORAGE_SITE_QUERY = "SELECT name, id, resourceid FROM inventory.storagesite";
+
+    private final TapClient<StorageSite> tapClient;
+    private final StorageSiteDAO storageSiteDAO;
+
+    /**
+     * Constructor that accepts a TapClient to query Storage Sites to sync.  Useful for testing.
+     *
+     * @param tapClient The TapClient to use.
+     */
+    public StorageSiteSync(final TapClient<StorageSite> tapClient, final StorageSiteDAO storageSiteDAO) {
+        this.tapClient = tapClient;
+        this.storageSiteDAO = storageSiteDAO;
+    }
+
+    /**
+     * Query for a Storage Site from a remote TAP (Luskan) service.  This method is expected to return a single
+     * StorageSite instance.  The result will typically be used to set to an Artifact's site locations during metadata
+     * synchronization.
+     *
+     * @return StorageSite  There should be a single StorageSite found.
+     * @throws AccessControlException permission denied
+     * @throws NotAuthenticatedException authentication attempt failed or rejected
+     * @throws ByteLimitExceededException input or output limit exceeded
+     * @throws IllegalArgumentException null method arguments or invalid query
+     * @throws ResourceNotFoundException remote resource not found
+     * @throws TransientException temporary failure of TAP service: same call could work in future
+     * @throws IOException failure to send or read data stream
+     * @throws InterruptedException thread interrupted
+     */
+    public StorageSite doit() throws AccessControlException, ResourceNotFoundException,
+                                     ByteLimitExceededException, NotAuthenticatedException,
+                                     IllegalArgumentException, TransientException, IOException,
+                                     InterruptedException {
+        try (final ResourceIterator<StorageSite> storageSiteIterator = queryStorageSites()) {
+            if (storageSiteIterator.hasNext()) {
+                final StorageSite storageSite = storageSiteIterator.next();
+                if (storageSiteIterator.hasNext()) {
+                    throw new IllegalStateException("More than one Storage Site found.");
+                } else {
```

Comment:
Before putting the StorageSite locally, you need to verify that the metaChecksum in the returned StorageSite's metaChecksum is correct by computing a new value and comparing to the assigned value. This method must fail on a mismatch.

the `storageSiteDAO.put()` and return do not need to be inside an else here

Reply:
This makes sense, thanks.